### DISCUSSION TITLE
Dynamically change empty mass and inertia in-flight

### DISCUSSION THREAD

**Discussion by scotteh432** [JSBSim-Team/jsbsim#961] [Q&A]
Hey everyone,

I'm attempting to model a parafoil in JSBSim, and have been presented with an interesting challenge. To accurately model a parafoil, it is necessary to be able to continuously change its mass and moment of inertia in flight, due to apparent mass effects acting on the parafoil by the air-mass. This is a common process done in parafoil dynamic modelling. Here is a link to a paper on the subject for the curious:

http://jtam.pl/Apparent-masses-and-inertia-moments-of-the-parafoil-,102141,0,2.html

This paper gives simple algebraic estimates apparent mass and moment of inertia terms. All scale as a function of air-density. These terms are very significant to the parafoil flight behaviour, especially for high-altitude deployments, where the parafoil inertia may be  half of what it is near-ground! This means I cannot simply include them as part of an unchanging empty-mass and empty-inertia properties if I want the parafoil to be modelled correctly through changing altitudes. In my JSBSim model, I have created functions of these apparent mass terms, but am unable to dynamically set the empty mass or empty moments of inertia terms in accordance with the changing apparent mass functions. These properties are read-only. Would anyone know of a way I can get around this limitation in JSBSim? **How can I dynamically change mass and moment of inertia?**

**What I've tried so far**

**1-Using fuel tanks:** tried positioning a fictious "fuel-tank" at the center of gravity and filled / emptied it with "fuel" dynamically with my apparent mass function. This worked but doing the equivalent with moments of inertia proved difficult. Changing all 3-axes moment of inertias in this way is a challenge, since adjusting the "fuel" mass in any tank offset from the CoG changes the inertia in 2 different axes. I tried deriving a relationship to calculate a "fuel" loading in 3 tanks (offset in X, Y, Z respectively), but this got complicated fast, so I gave up on this solution. There has to be an easier way.

**2-Using external forces:** Apply an external force / moment equal to impact of the mass / inertia. Like this: F_ap_mass = - app_mass * accel, in all three axes. And M_app_inert = - app_inert * rot_accel. I like the idea of this solution a lot, since the apparent mass effects technically act directionally (although there are simplifications to reduce them to one mass term I can use with reduced accuracy). This didn't work though; it would cause the simulation to become unstable. I'm not sure why. My theory is that the solver struggles with including acceleration terms in functions for forces, since accelerations should be calculated as a consequence of forces? I don't know. If anyone has any ideas as to why this is not working, I'm all ears.

Anyways, that's my problem. I need to dynamically change mass and moment of inertia to specific values calculated based on the flight condition. If anyone has any work-arounds to the empty-mass / empty-moments-of-inertia being read-only values problem, I'd love to hear them. Thank you so much for your help!

**Comment by seanmcleod** [comment]
Reminds me of something I brought up on the JSBSim mailing list in 2018.

-------
I noticed in both the Airbus A330 paper and the B747 paper that they model the moments of inertia and cg location in a slightly different manner to the normal JSBSim manner.

Instead of defining the empty weight moments of inertia and cg location and then making use of point masses to update the overall weight and update the moments of inertia and update the cg they define the moments of inertia using a formula or lookup table based on total weight and specify a cg location simply as cgx %.

So for example from the Airbus model:

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/8278d449-e145-4903-9495-c7b0264168ab)

Now looking at the source code for FGMassBalance it doesn’t look like it’s possible to set the moment of inertias using a function/table/property? It looks like you can only specify a fixed number and then you have to use point masses to effectively alter the moments of inertia by ‘adding them in’? 

And the same applies to the cg?

```xml
    <mass_balance>
        <ixx unit="{SLUG*FT2 | KG*M2}"> {number} </ixx>
        <iyy unit="{SLUG*FT2 | KG*M2}"> {number} </iyy>
        <izz unit="{SLUG*FT2 | KG*M2}"> {number} </izz>
        <ixy unit="{SLUG*FT2 | KG*M2}"> {number} </ixy>
        <ixz unit="{SLUG*FT2 | KG*M2}"> {number} </ixz>
        <iyz unit="{SLUG*FT2 | KG*M2}"> {number} </iyz>
        <emptywt unit="{LBS | KG"> {number} </emptywt>
        <location name="CG" unit="{IN | FT | M}">
            <x> {number} </x>
            <y> {number} </y>
            <z> {number} </z>
        </location>
```

Or have I missed a simple work-around/mapping that would allow one to easily implement something like the A330 model?

In particular in a Monte-Carlo setup the ideal would be to set a total weight property and cgx % property and have functions/tables that would work out the moments of inertia and cg location in the FDM.

-------

In the end no changes were made to the JSBSim source code to support the use of properties/functions/tables for setting/changing the moments of inertia and cg. It's still on my long list of JSBSim ToDo items to look into adding support to JSBSim to be able to easily and quickly make use of aircraft models that are specified as per the Airbus example above. 

Here are some snippets I'd also seen online in terms of how Boeing provides data to simulator manufacturers.

> 1.4 Mass Properties
The moments of inertia used in the simulator math model represent a typical combination of
passenger, cargo, and fuel. These nominal values are shown in figures 1.4.1 through 1.4.4.
For reference, the CG envelope is shown in figure 1.4.5.

**Roll Moment of Inertia - IXX** (ignore misspelling of weight)

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/fb314918-0b15-460a-bbf0-2d414375e46d)

**Comment by scotteh432** [comment]
Here's a few other use-cases off the top of my head for dynamic mass / inertias:

- Water-bombers
- Airships with slow-leaks
- Ablation of thermal protection on re-entry vehicles
- Ablative rocket engines
- and of course, apparent mass dynamics in light-weight wings (i.e. parafoils)

Instantaneous change can be modelled as a point-mass easily enough. But there are many cases where mass / inertia changes continuously and not due to fuel. Sometimes solutions can be hacked together using a combination fuel-tanks and point-masses. Other times not.

Did you ever find a work-around for your solution? Outside of a code-change to make mass / inertia be able to be dynamically updated? As much as I'd love the direct solution to be possible!

**Comment by bcoconni** [comment]
One option to modify the inertia without affecting the CG is to move 2 point masses that are located symmetrically with respect to the CG.

I have modified the script `scripts/ball_chute.xml` to exemplify that
```diff
   <description>Simplest of all tests</description>
   <use aircraft="ball" initialize="reset01"/>
   <run start="0.0" end="600" dt="0.008333">
+    <property value="5.0">inertia/pointmass-weight-lbs[0]</property>
+    <property value="5.0">inertia/pointmass-weight-lbs[1]</property>
+    <property value="0.0">inertia/pointmass-location-Y-inches[1]</property>
+
+    <event name="Change Y inertia">
+      <condition>
+        position/h-agl-ft lt 9000
+      </condition>
+      <set name="inertia/pointmass-location-Y-inches[0]" value="10.0" action="FG_STEP"/>
+      <set name="inertia/pointmass-location-Y-inches[1]" value="-10.0" action="FG_STEP"/>
+      <notify>
+        <property>inertia/weight-lbs</property>
+        <property>inertia/cg-x-in</property>
+        <property>inertia/cg-y-in</property>
+        <property>inertia/cg-x-in</property>
+        <property>inertia/ixx-slugs_ft2</property>
+        <property>inertia/iyy-slugs_ft2</property>
+        <property>inertia/izz-slugs_ft2</property>
+      </notify>
+    </event>
 
     <event name="Reef 1">
       <description> Test of parachute reefing </description>
@@ -15,6 +35,13 @@
       <set name="fcs/parachute_reef_pos_norm" value="0.2" action="FG_RAMP" tc="1.0"/>
       <notify>
         <property>simulation/sim-time-sec</property>
+        <property>inertia/weight-lbs</property>
+        <property>inertia/cg-x-in</property>
+        <property>inertia/cg-y-in</property>
+        <property>inertia/cg-x-in</property>
+        <property>inertia/ixx-slugs_ft2</property>
+        <property>inertia/iyy-slugs_ft2</property>
+        <property>inertia/izz-slugs_ft2</property>
       </notify>
     </event>
```

It gives the following output. As you can see the inertia Ixx and Iyy are modified by the CG position and Iyy are not.
```
Change Y inertia (Event 0) executed at time: 22.832420
    inertia/weight-lbs = 50.000000
    inertia/cg-x-in = 0.000000
    inertia/cg-y-in = 0.000000
    inertia/cg-x-in = 0.000000
    inertia/ixx-slugs_ft2 = 20.000000
    inertia/iyy-slugs_ft2 = 10.000000
    inertia/izz-slugs_ft2 = 10.000000


Reef 1 (Event 1) executed at time: 30.382118
    simulation/sim-time-sec = 30.382118
    inertia/weight-lbs = 50.000000
    inertia/cg-x-in = 0.000000
    inertia/cg-y-in = 0.000000
    inertia/cg-x-in = 0.000000
    inertia/ixx-slugs_ft2 = 20.215840
    inertia/iyy-slugs_ft2 = 10.000000
    inertia/izz-slugs_ft2 = 10.215840
```

**Comment by seanmcleod** [comment]
> Did you ever find a work-around for your solution? Outside of a code-change to make mass / inertia be able to be dynamically updated? 

No I didn't. At the time I think @bcoconni mentioned the "dumbbell" suggestion that he's mentioned above.

**Reply by bcoconni** [reply]
> At the time I think @bcoconni mentioned the "dumbbell" suggestion that he's mentioned above.

Oops ! Most likely I'm stuck in a time loop 😁

**Comment by scotteh432** [comment]
This is a good suggestion. I took it to it's logical conclusion, and added 6 point masses  (2 equal and opposite for each axis) and derived the relationship for controlling all 3 axes moments of inertia:

![free_body_diagram](https://github.com/JSBSim-Team/jsbsim/assets/25991545/6f50e602-4be5-4bb4-af4b-9a4c38d201af)

![derivation](https://github.com/JSBSim-Team/jsbsim/assets/25991545/41756b3b-f6ba-4ceb-a1f3-4dccc82cda15)

(My algebra skills are very fallible so if anyone wants to use this ^, I would recommend double checking my math)

There is an issue though. The system of equations has a imaginary roots if the added inertia (the C terms) in one axis is greater than the sum of the added inertia in the other two axes. The limitations of only being able to shift moment of inertia 2-axes at a time creates this. So it's not possible to shift to any arbitrary system of Ixx, Iyy and Izz, but there is a subset of Ixx, Iyy, Izz values where this solution works. 

I'll have to do some analysis work to figure out what situations work and which don't for the expected moment of inertia shifts for my system.

**Reply by bcoconni** [reply]
Hi @scotteh432,

Regarding your maths they look correct to me, if that proves anything 😄

The inequalities that you are obtaining are giving a volume which a 3 sides pyramid which apex is located at (0,0,0), see the blue surface in the figure below.

Your problem is that you have an inertia domain that you want to be able to cover (the orange surface in the figure below). Since the inertia domain is given by
$$C=\frac{I-I_{empty}}{2m}$$
you can always pick a mass $m$ that is sufficiently large so that the orange domain is entirely contained within the pyramid with the caveat that the sum of the "dumbbell" masses cannot exceed the total mass $M$ of the parafoil i.e. you must have $6m\le M$

![image](https://github.com/JSBSim-Team/jsbsim/assets/7502904/facefe92-6365-45d6-9e10-216550b691a4)

For the record, the Python code that I have used to generate the figure above:

```python
import numpy as np
import matplotlib.pyplot as plt

# Pyramid surface
h = np.arange(0,1.25,0.25)
Cx = np.outer([1,1,0,1], h)
Cy = np.outer([1,0,1,1], h)
Cz = np.outer([0,1,1,0], h)

# Dummy inertia surface (actually an ellipsoid)
r0 = 0.3
offset = 0.5
N = 20
u = np.linspace(0, 2 * np.pi, N)
v = np.linspace(0, np.pi, N)
x = r0 * np.outer(np.cos(u), np.sin(v))
y = r0 * np.outer(np.sin(u), np.sin(v))
z = 0.5*r0 * np.outer(np.ones(np.size(u)), np.cos(v))

fig = plt.figure()
ax=fig.add_subplot(projection='3d', computed_zorder=False)
ax.plot_surface(Cx,Cy,Cz)
ax.plot_surface(x+offset,y+offset,z+offset)
ax.set_xlabel('Cx')
ax.set_ylabel('Cy')
ax.set_zlabel('Cz')
ax.view_init(azim=20)
```

**Reply by bcoconni** [reply]
Some further thoughts: the figure above is wrong.

The orange solid must include the origin (0,0,0) which is basically the case when the vehicle inertia is equal to the "empty" inertia. So the correct representation of the problem is the figure below.

In most cases, the orange solid cannot be completely included within the pyramid whatever the choosen value of the "dumbbell" mass. The only feasible domain is the orange patch that is visible within the pyramid in the figure below:

![image](https://github.com/JSBSim-Team/jsbsim/assets/7502904/1e88c132-9e6a-4e6c-a837-f91e02e08d54)

The conclusion is that this problem can be only be addressed by adding a feature to JSBSim.

**Reply by scotteh432** [reply]
Thanks for the python visualization! I was able to take your code and make some modifications to include the parafoil inertial "requirements" with changing altitude to select an "empty" inertia that would take me through the range of inertias I need!

I'm not sure if this counts as a solution to this issue, but  it does give me what I need to move forward with my project. Crazy hack that this moving point mass solution is.

**Comment by scotteh432** [comment]
Update:

After messing about with some of the tutorial files, I appear to have found a solution that's much less of a hack. As it turns out, there is a JSBSim tutorial with this exact problem.  The airship model has a system file called "airship_added_mass.xml" . It's the same physical phenomenon in action (applied to airships rather than parafoils). To model the "mass" and "inertia" terms, they apply an external force / moment opposite to the direction of acceleration, whose magnitude is equal to the added mass / inertia. This has the same effect as adding additional mass or inertia: 

```
<fcs_function name="aero/added-mass/force-bx-lbs">
   <function>
    <product>
     <property> aero/added-mass/active </property>
     <property> atmosphere/rho-slugs_ft3 </property>
     <property> aero/constants/added-mass/k-axial </property>
     <property> aero/constants/volume-ft3 </property>
     <property> accelerations/udot-ft_sec2 </property>
    </product>
   </function>
   <output> external_reactions/added-mass-bx/magnitude </output>
  </fcs_function>
```

I fiddled with this a bit, and it does work, but seems to give me stability issues if the system is undergoing a lot of change (in my case, during parafoil deployment).  The airship solution to this problem is to just turn off the apparent mass effects until the system has stabilized after initialization. I implemented the same solution, and while it does work, I do lose the impacts of the apparent mass on parafoil inflation (something I've been struggling to model, and is a bulk of the point of my project). But this is probably the best I can hope for.

**Reply by tatjam** [reply]
I believe the problem with that approach is that you are using the previous time step acceleration, and thus not directly modifying the mass / inertia of the system. This is probably fine for slow changing systems but of course not appropriate for these kind of transients. (Although you could try to reduce the time step a lot and use a multi step integrator during deployment).
 
I'm working on a project that needs added mass modelling for airship simulation, and have met this exact problem. The solution is to implement the required changes directly into `FGAccelerations`. In our particular use case, this required implementing the full 6x6 added mass matrix used in airship modelling. Albeit a hit on performance (each time step you need to invert the mass matrix), this approach allows modelling pretty much everything, specially if you allow the mass matrix to evolve over time. 

Sadly, I'm unable to disclose the code right now, as it's only for internal use only. Maybe in the near future I'll be able to open a pull request with the modifications.

---

Nonetheless, the case exposed in the discussion probably doesn't need the full added mass matrix, and the modifications needed to JSBSim's code in that case are relatively easier. You could approach the problem by modifying `FGAccelerations` and / or `FGMassBalance`. In my opinion, modifying only `FGAccelerations` feels more appropriate as the added mass effects are kind of "external" to the vehicle. I'm going to quickly sketch a possible approach:

1) Add 10 `FGParameters` (1 for the added mass, 9 for the inertia terms. You could use less of course if you only need the diagonal terms!) to `FGAccelerations`
2) Implement the added mass effect here: https://github.com/JSBSim-Team/jsbsim/blob/86d5a6a2df987e3fc3319caa501ad8d9a9e30c9e/src/models/FGAccelerations.cpp#L188
Instead of division by `Mass`, divide by `Mass + mass_parameter.GetValue()`
3) (I'm not fully certain of this section*) Note that gravity, and other inertial forces, depend on the inertial mass of the object, and not its apparent mass including added masses. This means that the accelerations need to be multiplied by `Mass / (Mass + mass_parameter.GetValue())`. Thus, multiply this term here (the `vGravAccel` term) https://github.com/JSBSim-Team/jsbsim/blob/86d5a6a2df987e3fc3319caa501ad8d9a9e30c9e/src/models/FGAccelerations.cpp#L203
here https://github.com/JSBSim-Team/jsbsim/blob/86d5a6a2df987e3fc3319caa501ad8d9a9e30c9e/src/models/FGAccelerations.cpp#L202
and here, in the centrifugal term
https://github.com/JSBSim-Team/jsbsim/blob/86d5a6a2df987e3fc3319caa501ad8d9a9e30c9e/src/models/FGAccelerations.cpp#L193 
and finally in the Coriolis term:
https://github.com/JSBSim-Team/jsbsim/blob/86d5a6a2df987e3fc3319caa501ad8d9a9e30c9e/src/models/FGAccelerations.cpp#L190
although I'm not certain if this is correct (see the footnote)
3) Implement the added inertia effect in this function: https://github.com/JSBSim-Team/jsbsim/blob/86d5a6a2df987e3fc3319caa501ad8d9a9e30c9e/src/models/FGAccelerations.cpp#L138
You would create a new `FGMatrix33` and add it to the existing `in.J`. Of course this also requires inversion to calculate the new `in.Jinv`, but 3x3 inversion is extremely fast (and already implemented in JSBSim!)
4) Expose the parameters with desired names, and implement the script to exploit them

---

(*) I presume this is the case, as the added mass effect was discovered investigating pendulums, and in that case the gravity "appears to diminish". I think the real justification is that the fluid is also accelerated by gravity, but it has already reached equilibrium. Note that this doesn't violate Einstein equivalence principle, if you assume you are in some massive box, accelerating "upwards", the fluid would also eventually reach equilibrium and the situation would be impossible to tell apart from gravity. 😉

This same reasoning can be applied to the centrifugal term, but not really to the Coriolis term, though the effects are probably too small to care about in this kind of model.
